src/Core/3DTiles/C3DTileset.js

```diff
@@ -0,0 +1,112 @@
+        this.extensionsUsed = json.extensionsUsed;
+        this.extensionsRequired = json.extensionsRequired;
+        this.tiles = [];
+        this.parseTiles(json.root, baseURL, undefined, registeredExtensions);
+
+        if (json.extensions) {
+            this.extensions =
+                registeredExtensions.parseExtensions(json.extensions, this);
+        }
+    }
+
+    /**
+     * Recursion on the 3DTiles tileset (which is a tree) to parse the tiles
+     * (nodes of the tree).
+     * @param {object} tile - current tile
+     * @param {string} baseURL - url of the source tileset and tiles
+     * @param {object} parent - parent tile (used for computing the transform
+     * matrix from local to global coordinates)
+     * @param {object} registeredExtensions - 3D Tiles extensions registered
+     * in the C3DTilesLayer (see {@link C3DTExtensions}
+     */
+    parseTiles(tile, baseURL, parent, registeredExtensions) {
+        const inverseTileTransform = new THREE.Matrix4();
+
+        // compute transform (will become Object3D.matrix when the object is
+        // downloaded)
+        tile.transform =
+            tile.transform ? (new THREE.Matrix4()).fromArray(tile.transform) :
+                undefined;
+
+        // The only reason to store _worldFromLocalTransform is because of
+        // extendTileset where we need the transform chain for one tile.
+        tile._worldFromLocalTransform = tile.transform;
+        if (parent && parent._worldFromLocalTransform) {
+            if (tile.transform) {
+                tile._worldFromLocalTransform =
+                    new THREE.Matrix4().multiplyMatrices(
+                        parent._worldFromLocalTransform, tile.transform);
+            } else {
+                tile._worldFromLocalTransform = parent._worldFromLocalTransform;
+            }
+        }
+
+        // inverseTileTransform is only used for volume.region
+        if ((tile.viewerRequestVolume && tile.viewerRequestVolume.region)
+            || (tile.boundingVolume && tile.boundingVolume.region)) {
+            if (tile._worldFromLocalTransform) {
+                inverseTileTransform.getInverse(tile._worldFromLocalTransform);
+            } else {
+                inverseTileTransform.identity();
+            }
+        }
+
+        tile.viewerRequestVolume = tile.viewerRequestVolume ?
+            new C3DTBoundingVolume(tile.viewerRequestVolume,
+                inverseTileTransform,
+                registeredExtensions) : undefined;
+        tile.boundingVolume = tile.boundingVolume ?
+            new C3DTBoundingVolume(tile.boundingVolume,
+                inverseTileTransform, registeredExtensions) : undefined;
+
+        this.tiles.push(tile);
+        tile.tileId = this.tiles.length - 1;
+        tile.baseURL = baseURL;
+        if (tile.children) {
+            for (const child of tile.children) {
+                this.parseTiles(child, baseURL, tile, registeredExtensions);
```

Comment:
I notice that the tree structure is flattened and that this is also the case in the master branch.

here:
```js
this.parseTiles(child, baseURL, tile, registeredExtensions);
```

All `child` and `tile`  will finally be added in `this.tiles`, is that normal?

Reply:
Yes it is already the case on the master branch. In addition to that, there is a tree structure with the nodes representing the tiles contents stored as threejs objects in the `object3d` attribute of the layer. (it is in fact a bit more complicated than this, see issue #1039 )

I don't know why this choice has been made in the first place but I think we may be able to simplify this and keep only one tree structure with the tiles and their content loaded as threejs objects (as discussed in the third item of issue #1171 ). I haven't looked into it yet though and I think it might be better to do it in a future PR ?

Reply:
Yes, indeed a major refactor is necessary.